File: android/src/main/kotlin/com/rationalowl/rationalowl_flutter/FlutterRationalOwlPlugin.kt
package com.rationalowl.rationalowl_flutter

import com.rationalowl.minerva.client.android.DeviceRegisterResultListener
import com.rationalowl.minerva.client.android.MessageListener
import com.rationalowl.minerva.client.android.MinervaManager
import com.rationalowl.minerva.client.android.util.Logger
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.json.JSONObject
import java.lang.reflect.Field
import kotlin.collections.ArrayList

private const val CHANNEL_NAME = "plugins.rationalowl.com/rationalowl_flutter"

class FlutterRationalOwlPlugin : FlutterPlugin, MethodCallHandler, DeviceRegisterResultListener,
    MessageListener {
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)

        MinervaManager.init(binding.applicationContext)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        var value: Any? = null

        when (call.method) {
            "MinervaManager#setRegisterResultListener" -> setRegisterResultListener()
            "MinervaManager#clearRegisterResultListener" -> clearRegisterResultListener()
            "MinervaManager#setMsgListener" -> setMsgListener()
            "MinervaManager#clearMsgListener" -> clearMsgListener()
            "MinervaManager#becomeActive" -> becomeActive()
            "MinervaManager#enterBackground" -> enterBackground()
            "MinervaManager#setDeviceToken" -> setDeviceToken(call.arguments())
            "MinervaManager#enableNotificationTracking" -> enableNotificationTracking(call.arguments())
            "MinervaManager#registerDevice" -> registerDevice(call.arguments())
            "MinervaManager#unregisterDevice" -> unregisterDevice(call.arguments())
            "MinervaManager#isDeviceAppRegister" -> value = isDeviceAppRegister(call.arguments())
            "MinervaManager#sendUpstreamMsg" -> value = sendUpstreamMsg(call.arguments())
            "MinervaManager#sendP2PMsg" -> value = sendP2PMsg(call.arguments())
            else -> {
                result.notImplemented()
                return
            }
        }

        result.success(value)
    }

    private fun setRegisterResultListener() {
        val minMgr = MinervaManager.getInstance()
        minMgr.setRegisterResultListener(this)
    }

    private fun clearRegisterResultListener() {
        val minMgr = MinervaManager.getInstance()
        minMgr.clearRegisterResultListener()
    }

    private fun setMsgListener() {
        val minMgr = MinervaManager.getInstance()
        minMgr.setMsgListener(this)
    }

    private fun clearMsgListener() {
        val minMgr = MinervaManager.getInstance()
        minMgr.clearMsgListener()
    }

    private fun becomeActive() {
        val minMgr = MinervaManager.getInstance()
        minMgr.becomeActive()
    }

    private fun enterBackground() {
        val minMgr = MinervaManager.getInstance()
        minMgr.enterBackground()
    }

    private fun enableNotificationTracking(arguments: Map<String, Any>?) {
        try {
            println("enableNotificationTracking started")
            val minMgr = MinervaManager.getInstance()
            val data = arguments?.get("data") as Map<String, String>
            minMgr.enableNotificationTracking(data)
            println("enableNotificationTracking done")
        } catch (e: Exception) {
            println(e);
        }
    }

    private fun setDeviceToken(arguments: Map<String, Any>?) {
        val minMgr = MinervaManager.getInstance()
        val deviceToken = arguments?.get("deviceToken") as String
        minMgr.setDeviceToken(deviceToken)
    }

    private fun registerDevice(arguments: Map<String, Any>?) {
        val minMgr = MinervaManager.getInstance()
        val gateHost = arguments?.get("gateHost") as String
        val serviceId = arguments["serviceId"] as String
        val deviceRegName = arguments["deviceRegName"] as String
        minMgr.registerDevice(gateHost, serviceId, deviceRegName)
    }

    private fun unregisterDevice(arguments: Map<String, Any>?) {
        val minMgr = MinervaManager.getInstance()
        val serviceId = arguments?.get("serviceId") as String
        minMgr.unregisterDevice(serviceId)
    }

    private fun isDeviceAppRegister(arguments: Map<String, Any>?): Boolean {
        val minMgr = MinervaManager.getInstance()
        val serviceId = arguments?.get("serviceId") as String
        val deviceRegId = arguments["deviceRegId"] as String
        return minMgr.isDeviceAppRegister(serviceId, deviceRegId)
    }

    private fun sendUpstreamMsg(arguments: Map<String, Any>?): String? {
        val minMgr = MinervaManager.getInstance()
        val data = arguments?.get("data") as String
        val serverRegId = arguments["serverRegId"] as String
        return minMgr.sendUpstreamMsg(data, serverRegId)
    }

    private fun sendP2PMsg(arguments: Map<String, Any>?): String? {
        val minMgr = MinervaManager.getInstance()
        val data = arguments?.get("data") as String
        val destDevices = arguments.get("destDevices") as ArrayList<String>
        val supportMsgQ = arguments["supportMsgQ"] as Boolean
        val notiTitle = arguments["notiTitle"] as String?
        val notiBody = arguments["notiBody"] as String?
        return minMgr.sendP2PMsg(data, destDevices, supportMsgQ, notiTitle, notiBody)
    }

    public override fun onRegisterResult(
        resultCode: Int,
        resultMsg: String?,
        deviceRegId: String?
    ) {
        channel.invokeMethod(
            "DeviceRegisterResultListener#onRegisterResult", mapOf(
                "resultCode" to resultCode,
                "resultMsg" to resultMsg,
                "deviceRegId" to deviceRegId
            )
        )
    }

    public override fun onUnregisterResult(resultCode: Int, resultMsg: String?) {
        channel.invokeMethod(
            "DeviceRegisterResultListener#onUnregisterResult", mapOf(
                "resultCode" to resultCode,
                "resultMsg" to resultMsg,
            )
        )
    }

    private fun decodeMsgList(msgList: ArrayList<JSONObject>): List<Map<String, Any?>> {
        return msgList.map {
            val data = it.getString("data")
            it.put("data", JSONObject(data).toMap())
            it.toMap()
        }
    }

    public override fun onDownstreamMsgRecieved(msgList: ArrayList<JSONObject>) {
        channel.invokeMethod("MessageListener#onDownstreamMsgRecieved", decodeMsgList(msgList))
    }

    public override fun onP2PMsgRecieved(msgList: ArrayList<JSONObject>) {
        channel.invokeMethod("MessageListener#onP2PMsgRecieved", decodeMsgList(msgList))
    }

    public override fun onPushMsgRecieved(msgList: ArrayList<JSONObject>) {
        channel.invokeMethod("MessageListener#onPushMsgRecieved", decodeMsgList(msgList))
    }

    public override fun onSendUpstreamMsgResult(
        resultCode: Int,
        resultMsg: String?,
        msgId: String?
    ) {
        channel.invokeMethod(
            "MessageListener#onSendUpstreamMsgResult", mapOf(
                "resultCode" to resultCode,
                "resultMsg" to resultMsg,
                "msgId" to msgId
            )
        )
    }

    public override fun onSendP2PMsgResult(
        resultCode: Int,
        resultMsg: String?,
        msgId: String?
    ) {
        channel.invokeMethod(
            "MessageListener#onSendP2PMsgResult", mapOf(
                "resultCode" to resultCode,
                "resultMsg" to resultMsg,
                "msgId" to msgId
            )
        )
    }

    private fun JSONObject.toMap(): Map<String, Any?> {
        val map = HashMap<String, Any?>()
        keys().forEach { key -> map[key] = this[key] }

        return map
    }
}

File: ios/Classes/FlutterRationalOwlPlugin.swift
import Flutter
import UIKit

import RationalOwl

let channelName = "plugins.rationalowl.com/rationalowl_flutter"

public class FlutterRationalOwlPlugin: NSObject, FlutterPlugin {
    private static var channel: FlutterMethodChannel!

    private weak var originalNotificationCenterDelegate: UNUserNotificationCenterDelegate?

    private struct RespondsTo {
        var willPresentNotification: Bool = false
        var didReceiveNotificationResponse: Bool = false
        var openSettingsFor: Bool = false
    }

    private var originalNotificationCenterDelegateRespondsTo = RespondsTo()

    public static func register(with registrar: FlutterPluginRegistrar) {
        channel = FlutterMethodChannel(name: channelName, binaryMessenger: registrar.messenger())

        let instance = FlutterRationalOwlPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addApplicationDelegate(instance)
    }

    override private init() {
        super.init()
        NotificationCenter.default.addObserver(self, selector: #selector(application_onDidFinishLaunchingNotification(_:)), name: UIApplication.didFinishLaunchingNotification, object: nil)
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        var value: Any? = nil

        switch call.method {
        case "MinervaManager#setRegisterResultListener":
            setRegisterResultListener()
        case "MinervaManager#clearRegisterResultListener":
            clearRegisterResultListener()
        case "MinervaManager#setMsgListener":
            setMsgListener()
        case "MinervaManager#clearMsgListener":
            clearMsgListener()
        case "MinervaManager#becomeActive":
            becomeActive()
        case "MinervaManager#enterBackground":
            enterBackground()
        case "MinervaManager#setDeviceToken":
            setDeviceToken(call.arguments as! [String: Any])
        case "MinervaManager#setAppGroup":
            setAppGroup(call.arguments as! [String: Any])
        case "MinervaManager#enableNotificationTracking":
            enableNotificationTracking(call.arguments as! [String: Any])
        case "MinervaManager#registerDevice":
            registerDevice(call.arguments as! [String: Any])
        case "MinervaManager#unregisterDevice":
            unregisterDevice(call.arguments as! [String: Any])
        case "MinervaManager#isDeviceAppRegister":
            value = isDeviceAppRegister(call.arguments as! [String: Any])
        case "MinervaManager#sendUpstreamMsg":
            value = sendUpstreamMsg(call.arguments as! [String: Any])
        case "MinervaManager#sendP2PMsg":
            value = sendP2PMsg(call.arguments as! [String: Any])
        default:
            result(FlutterMethodNotImplemented)
            return
        }

        result(value)
    }

    private func setRegisterResultListener() {
        let minMgr = MinervaManager.getInstance()!
        minMgr.setDeviceRegisterResultDelegate(self)
    }

    private func clearRegisterResultListener() {
        let minMgr = MinervaManager.getInstance()!
        minMgr.clearDeviceRegisterResultDelegate()
    }

    private func setMsgListener() {
        let minMgr = MinervaManager.getInstance()!
        minMgr.setMessageDelegate(self)
    }

    private func clearMsgListener() {
        let minMgr = MinervaManager.getInstance()!
        minMgr.clearMessageDelegate()
    }

    private func becomeActive() {
        let minMgr = MinervaManager.getInstance()!
        minMgr.becomeActive()
    }

    private func enterBackground() {
        let minMgr = MinervaManager.getInstance()!
        minMgr.enterBackground()
    }

    private func enableNotificationTracking(_ arguments: [String: Any]) {
        let minMgr = MinervaManager.getInstance()!
        let data = arguments["data"] as! [AnyHashable: Any]
        let appGroup = arguments["appGroup"] as! String
        minMgr.enableNotificationTracking(data, appGroup: appGroup)
    }

    private func setDeviceToken(_ arguments: [String: Any]) {
        let minMgr = MinervaManager.getInstance()!
        let deviceToken = arguments["deviceToken"] as! String
        minMgr.setDeviceToken(deviceToken)
    }

    private func setAppGroup(_ arguments: [String: Any]) {
        let minMgr = MinervaManager.getInstance()!
        let appGroup = arguments["appGroup"] as! String
        minMgr.setAppGroup(appGroup)
    }

    private func registerDevice(_ arguments: [String: Any]) {
        let minMgr = MinervaManager.getInstance()!
        let gateHost = arguments["gateHost"] as! String
        let serviceId = arguments["serviceId"] as! String
        let deviceRegName = arguments["deviceRegName"] as! String
        minMgr.registerDevice(gateHost, serviceId: serviceId, deviceRegName: deviceRegName)
    }

    private func unregisterDevice(_ arguments: [String: Any]) {
        let minMgr = MinervaManager.getInstance()!
        let serviceId = arguments["serviceId"] as! String
        minMgr.unregisterDevice(serviceId)
    }

    private func isDeviceAppRegister(_ arguments: [String: Any]) -> Bool {
        let minMgr = MinervaManager.getInstance()!
        let serviceId = arguments["serviceId"] as! String
        let deviceRegId = arguments["deviceRegId"] as! String
        return minMgr.isDeviceAppRegister(serviceId, deviceRegId: deviceRegId)
    }

    private func sendUpstreamMsg(_ arguments: [String: Any]) -> String? {
        let minMgr = MinervaManager.getInstance()!
        let data = arguments["data"] as! String
        let serverRegId = arguments["serverRegId"] as! String
        return minMgr.sendUpstreamMsg(data, serverRegId: serverRegId)
    }

    private func sendP2PMsg(_ arguments: [String: Any]) -> String? {
        let minMgr = MinervaManager.getInstance()!
        let data = arguments["data"] as! String
        let destDevices = arguments["destDevices"] as! [String]
        let supportMsgQ = arguments["supportMsgQ"] as! Bool
        let notiTitle = arguments["notiTitle"] as? String
        let notiBody = arguments["notiBody"] as? String
        return minMgr.sendP2PMsg(data, devices: destDevices, supportMsgQ: supportMsgQ, notiTitle: notiTitle, notiBody: notiBody)
    }

    @objc func application_onDidFinishLaunchingNotification(_ notification: Notification) {
        var shouldReplaceDelegate = true
        let notificationCenter = UNUserNotificationCenter.current()

        if notificationCenter.delegate != nil {
            if notificationCenter.delegate is FlutterAppLifeCycleProvider {
                shouldReplaceDelegate = false
            }

            if shouldReplaceDelegate {
                originalNotificationCenterDelegate = notificationCenter.delegate
                originalNotificationCenterDelegateRespondsTo.openSettingsFor = originalNotificationCenterDelegate!.responds(to: #selector(UNUserNotificationCenterDelegate.userNotificationCenter(_:openSettingsFor:)))
                originalNotificationCenterDelegateRespondsTo.willPresentNotification = originalNotificationCenterDelegate!.responds(to: #selector(UNUserNotificationCenterDelegate.userNotificationCenter(_:willPresent:withCompletionHandler:)))
                originalNotificationCenterDelegateRespondsTo.didReceiveNotificationResponse = originalNotificationCenterDelegate!.responds(to: #selector(UNUserNotificationCenterDelegate.userNotificationCenter(_:didReceive:withCompletionHandler:)))
            }
        }

        if shouldReplaceDelegate {
            notificationCenter.delegate = self
        }
    }
}

extension FlutterRationalOwlPlugin: UIApplicationDelegate {
    public func applicationDidBecomeActive(_ application: UIApplication) {
        let minMgr = MinervaManager.getInstance()!
        minMgr.becomeActive()
    }

    public func applicationWillResignActive(_ application: UIApplication) {
        let minMgr = MinervaManager.getInstance()!
        minMgr.enterBackground()
    }
}

extension FlutterRationalOwlPlugin: UNUserNotificationCenterDelegate {
    public func userNotificationCenter(_ center: UNUserNotificationCenter,
                                       didReceive response: UNNotificationResponse,
                                       withCompletionHandler completionHandler: @escaping () -> Void)
    {
        let userInfo = response.notification.request.content.userInfo
        let minMgr = MinervaManager.getInstance()!
        minMgr.receivedApns(userInfo)

        if originalNotificationCenterDelegate != nil && originalNotificationCenterDelegateRespondsTo.didReceiveNotificationResponse {
            originalNotificationCenterDelegate!.userNotificationCenter!(center, didReceive: response, withCompletionHandler: completionHandler)
        } else {
            completionHandler()
        }
    }

    public func userNotificationCenter(_ center: UNUserNotificationCenter,
                                       willPresent notification: UNNotification,
                                       withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void)
    {
        if originalNotificationCenterDelegate != nil && originalNotificationCenterDelegateRespondsTo.willPresentNotification {
            originalNotificationCenterDelegate!.userNotificationCenter!(center, willPresent: notification, withCompletionHandler: completionHandler)
        } else {
            completionHandler([])
        }
    }

    public func userNotificationCenter(_ center: UNUserNotificationCenter,
                                       openSettingsFor notification: UNNotification?)
    {
        if originalNotificationCenterDelegate != nil && originalNotificationCenterDelegateRespondsTo.openSettingsFor {
            originalNotificationCenterDelegate!.userNotificationCenter!(center, openSettingsFor: notification)
        }
    }
}

extension FlutterRationalOwlPlugin: DeviceRegisterResultDelegate {
    public func onRegisterResult(_ resultCode: Int32, resultMsg: String?, deviceRegId: String?) {
        var args: [String: Any] = [
            "resultCode": resultCode
        ]

        if resultMsg != nil {
            args["resultMsg"] = resultMsg
        }

        if deviceRegId != nil {
            args["deviceRegId"] = deviceRegId
        }

        DispatchQueue.main.async {
            Self.channel.invokeMethod("DeviceRegisterResultListener#onRegisterResult", arguments: args)
        }
    }

    public func onUnregisterResult(_ resultCode: Int32, resultMsg: String?) {
        var args: [String: Any] = [
            "resultCode": resultCode
        ]

        if resultMsg != nil {
            args["resultMsg"] = resultMsg
        }

        DispatchQueue.main.async {
            Self.channel.invokeMethod(
                "DeviceRegisterResultListener#onUnregisterResult", arguments: args
            )
        }
    }
}

extension FlutterRationalOwlPlugin: MessageDelegate {
    public func onDownstreamMsgRecieved(_ msgList: [Any]!) {
        DispatchQueue.main.async {
            Self.channel.invokeMethod("MessageListener#onDownstreamMsgRecieved", arguments: msgList)
        }
    }

    public func onP2PMsgRecieved(_ msgList: [Any]!) {
        let decodedMsgList = msgList.map { msg in
            guard var msgDict = msg as? [String: Any] else { return msg }

            guard let dataString = msgDict["data"] as? String,
                  let data = dataString.data(using: .utf8) else { return msgDict }

            do {
                msgDict["data"] = try JSONSerialization.jsonObject(with: data)
            } catch {}

            return msgDict
        }

        DispatchQueue.main.async {
            Self.channel.invokeMethod("MessageListener#onP2PMsgRecieved", arguments: decodedMsgList)
        }
    }

    public func onPushMsgRecieved(_ msgSize: Int32, msgList: [Any]!, alarmIdx: Int32) {
        DispatchQueue.main.async {
            Self.channel.invokeMethod("MessageListener#onPushMsgRecieved", arguments: msgList)
        }
    }

    public func onUpstreamMsgResult(_ resultCode: Int32, resultMsg: String?, msgId: String?) {
        var args: [String: Any] = [
            "resultCode": resultCode
        ]

        if resultMsg != nil {
            args["resultMsg"] = resultMsg
        }

        if msgId != nil {
            args["msgId"] = msgId
        }

        DispatchQueue.main.async {
            Self.channel.invokeMethod("MessageListener#onSendUpstreamMsgResult", arguments: args)
        }
    }

    public func onP2PMsgResult(_ resultCode: Int32, resultMsg: String?, msgId: String?) {
        var args: [String: Any] = [
            "resultCode": resultCode
        ]

        if resultMsg != nil {
            args["resultMsg"] = resultMsg
        }

        if msgId != nil {
            args["msgId"] = msgId
        }

        DispatchQueue.main.async {
            Self.channel.invokeMethod("MessageListener#onSendP2PMsgResult", arguments: args)
        }
    }
}
